changeset_test.exs

```diff
@@ -0,0 +1,516 @@
+        |> Changeset.new(%{name: "foo"})
+        |> Api.create!()
+
+      {:ok, %{category: category}}
+    end
+
+    test "it get an attribute of the changeset", %{category: category} do
+      changeset =
+        category
+        |> Changeset.new(%{name: "bar"})
+
+      assert "bar" == Changeset.get_attribute(changeset, :name)
+    end
+
+    test "it falls back to the original value of the attribute", %{category: category} do
+      changeset =
+        category
+        |> Changeset.new()
+
+      assert "foo" == Changeset.get_attribute(changeset, :name)
+    end
+
+    test "it returns nil if the attribute not found" do
+      changeset =
+        Category
+        |> Changeset.new()
+
+      assert %{} == changeset.attributes
+
+      assert is_nil(Changeset.get_attribute(changeset, :not_there))
+    end
+  end
+
+  describe "fetch_change/2" do
+    test "it get a changed attribute of the changeset" do
+      category =
+        Category
+        |> Changeset.new(%{name: "foo"})
+        |> Api.create!()
+
+      changeset =
+        category
+        |> Changeset.new(%{name: "bar"})
+
+      assert {:ok, "bar"} == Changeset.fetch_change(changeset, :name)
+    end
+
+    test "it returns :error if the attribute not found" do
+      changeset =
+        Category
+        |> Changeset.new()
+
+      assert %{} == changeset.attributes
+
+      assert :error == Changeset.fetch_change(changeset, :none)
+    end
+  end
+
+  describe "set_context/2" do
+    test "it sets context to a given map" do
+      changeset =
+        Category
+        |> Changeset.new()
+        |> Changeset.set_context(%{key: "value"})
+
+      assert "value" == changeset.context.key
+    end
+  end
+
+  describe "put_context/2" do
+    test "it puts a value in a context key" do
+      changeset =
+        Category
+        |> Changeset.new()
+        |> Changeset.put_context(:key, "value")
+
+      assert "value" == changeset.context.key
+    end
+  end
+
+  describe "append_to_relationship/3" do
+    test "it appends entities to a resource's relationship" do
+      post1 = Post |> Changeset.new(%{title: "title1"}) |> Api.create!()
+      post2 = Post |> Changeset.new(%{title: "title2"}) |> Api.create!()
+
+      changeset =
+        Author
+        |> Changeset.new()
+        |> Changeset.append_to_relationship(:posts, [post1, post2])
+
+      assert Enum.sort([%{id: post1.id}, %{id: post2.id}]) ==
+               Enum.sort(changeset.relationships.posts.add)
+    end
+
+    test "it accepts value of single attribute primary_key as a second param" do
+      post1 = Post |> Changeset.new(%{title: "title1"}) |> Api.create!()
+
+      changeset =
+        Author
+        |> Changeset.new()
+        |> Changeset.append_to_relationship(:posts, post1.id)
+
+      assert %{add: [%{id: post1.id}]} == changeset.relationships.posts
+    end
+
+    test "it accepts a map %{id: value} representing primary key as a second param only if primary key is a single attribute" do
```

Comment:
this works for composite primary keys as well.

Reply:
I could not catch this case
I tried to define 2 attributes of a resource as primary keys like 
``` elixir
attributes do
      attribute :id, :uuid, primary_key?: true, default: &Ecto.UUID.generate/0
      attribute :title, :string, primary_key?: true
      attribute :contents, :string
    end
```
but it did not work, this line will always match https://github.com/kingshalaby1/ash/blob/ae5ff031a0da0f9758714ceb85b2237c59f3d05c/lib/ash/changeset/changeset.ex#L637
which consequently ends up with an error 
Am I trying a wrong way, or is there another way to define composite primary key?

Reply:
Hmm...if you could add that resource w/ a composite key, and write the failing test, I can check it out.

Reply:
> Awesome, I appreciate you adding that test. Go ahead and comment it out for now, and we can fix it later. It may be a good task for you to take?
yes, I'd love to